### PR TITLE
compiler: Uncache data carriers

### PR DESCRIPTION
**Pull request by FabioLuporini**
Scalar data carriers are unique, so why attempt to cache them?

### CODE REVIEW

File: examples/performance/01_gpu.ipynb [hunk 1]
```diff
@@ -303,7 +303,7 @@
       "      for (int y = y_m; y <= y_M; y += 1)\n",
       "      {\n",
       "        float r3 = -2.0F*u[time][x + 2][y + 2];\n",
-      "        u[time + 1][x + 2][y + 2] = dt*(r0*u[time][x + 2][y + 2] + c*(r1*r3 + r1*u[time][x + 1][y + 2] + r1*u[time][x + 3][y + 2] + r2*r3 + r2*u[time][x + 2][y + 1] + r2*u[time][x + 2][y + 3]));\n",
+      "        u[time + 1][x + 2][y + 2] = dt*(c*(r1*r3 + r1*u[time][x + 1][y + 2] + r1*u[time][x + 3][y + 2] + r2*r3 + r2*u[time][x + 2][y + 1] + r2*u[time][x + 2][y + 3]) + r0*u[time][x + 2][y + 2]);\n",
```
Comment (georgebisbas) [inline examples/performance/01_gpu.ipynb:306]: Is this change of order deterministic?

**Review by mloubout — COMMENTED** [review]
Looks gtg.  Not sue why the symbol order changes from ascending to descending, I guess `id` 's related.

**Comment by FabioLuporini** [comment]
@mloubout 

> Looks gtg. Not sue why the symbol order changes from ascending to descending, I guess id 's related.

Oh, it boils down to how SymPy internally orders objects... it uses the `cls.__name__` , and this PR is altering the class hierarchy, so...

File: tests/test_caching.py [hunk 2]
```diff
@@ -168,6 +170,28 @@ def test_bound_symbol(self):
...
+        dtype = type('Bar', (c_void_p,), {})
+
+        foo0 = Object('foo', dtype, v0)
+        foo1 = Object('foo', dtype, v0)
+        foo2 = Object('foo', dtype, v1)
+
+        # Obviously:
+        assert foo0 is not foo1
+        assert foo0 is not foo2
+        assert foo1 is not foo2
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply tests/test_caching.py:186]: "checked" how? like `foo0.dtype is foo1.dtype` ? that's obvious imho

**Review by mloubout — APPROVED** [review]

**Pull request merged**